ipld/retriever.go

```diff
@@ -106,47 +127,67 @@ func (r *Retriever) newSession(ctx context.Context, dah *da.DataAvailabilityHead
 	}
 
 	ses.squareImported = square
+	go ses.request(ctx)
 	return ses, nil
 }
 
-func (rs *retrieverSession) retrieve(ctx context.Context, q *quadrant) (*rsmt2d.ExtendedDataSquare, error) {
-	defer func() {
-		// all shares which were requested or repaired are written to disk with the commit
-		// we store *all*, so they are served to the network, including incorrectly committed data(BEFP case),
-		// so that network can check BEFP
-		if err := rs.adder.Commit(); err != nil {
-			log.Errorw("committing DAG", "err", err)
-		}
-	}()
-	// request quadrant and fill it into rs.square slice
-	// we don't care about the errors here, just need to request as much data as we can to be able to reconstruct below
-	rs.request(ctx, q)
+// Done notifies user to try reconstruction.
+// Try because there is no way currently to guarantedd when
+func (rs *retrievalSession) Done() <-chan struct{} {
+	return rs.done
+}
 
-	// try repair
-	// TODO: Avoid reimporting of the square which can potentially remove the requirement for the lock
+// Reconstruct tries to reconstruct the data square and returns it on success.
+func (rs *retrievalSession) Reconstruct() (*rsmt2d.ExtendedDataSquare, error) {
+	// prevent further writes to the square
 	rs.squareLk.Lock()
 	defer rs.squareLk.Unlock()
+	// and try to repair with what we have
 	err := rs.squareImported.Repair(rs.dah.RowsRoots, rs.dah.ColumnRoots, rs.codec, rs.treeFn)
 	if err != nil {
 		return nil, err
 	}
-
-	log.Infow("data square reconstructed", "data_hash", hex.EncodeToString(rs.dah.Hash()), "size", len(rs.dah.RowsRoots))
+	log.Debugw("data square reconstructed", "data_hash", hex.EncodeToString(rs.dah.Hash()), "size", len(rs.dah.RowsRoots))
 	return rs.squareImported, nil
 }
 
-func (rs *retrieverSession) request(ctx context.Context, q *quadrant) {
-	ctx, cancel := context.WithTimeout(ctx, RetrieveQuadrantTimeout)
-	defer cancel()
+func (rs *retrievalSession) Close() error {
+	// all shares which were requested or repaired are written to disk with the commit
+	// we store *all*, so they are served to the network, including incorrectly committed
+	// data(BEFP case), so that network can check BEFP
```

Comment:
```suggestion
	// Note that we store *all*, so they are served to the network, including incorrectly committed
	// data(BEFP case), so that the network can check BEFP.
```